### PR TITLE
Fix to #34895 - Final GroupBy on an entity type adds orderings for all of its properties

### LINKED ISSUE #34895: Final GroupBy on an entity type adds orderings for all of its properties
The following final GroupBy over an entity type:

```c#
_ = await context.Posts.GroupBy(p => p.Blog).ToListAsync();
```

... results in the following SQL:

```sql
SELECT [b].[Id], [b].[Author], [b].[Name], [p].[Id], [p].[BlogId], [p].[Title]
FROM [Posts] AS [p]
INNER JOIN [Blogs] AS [b] ON [p].[BlogId] = [b].[Id]
ORDER BY [b].[Id], [b].[Author], [b].[Name]
```

While it's understandable that we order by the primary key columns (so as to have all Posts of the same Blog sorted together, for streaming materialization<sup>*</sup>), I don't think we should be ordering by the other Blog properties (Author, Name). This both requires extra work from the database, and also fails if Blog has a column which isn't sortable (see https://github.com/npgsql/efcore.pg/issues/3202).

<sup>*</sup> Note: #29171 tracks removing the ordering for the primary key properties as well, removing that burden from the database.

Originally flagged by @LipatovAlexander in https://github.com/npgsql/efcore.pg/issues/3202

<details>
<summary>Complete repro</summary>

```c#
await using var context = new BlogContext();
await context.Database.EnsureDeletedAsync();
await context.Database.EnsureCreatedAsync();

_ = await context.Posts.GroupBy(p => p.Blog).ToListAsync();

public class BlogContext : DbContext
{
    public DbSet<Blog> Blogs { get; set; }
    public DbSet<Post> Posts { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseSqlServer("Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0;Encrypt=false")
            .LogTo(Console.WriteLine, LogLevel.Information)
            .EnableSensitiveDataLogging();
}

public class Blog
{
    public int Id { get; set; }
    public string Name { get; set; }
    public string Author { get; set; }

    public List<Post> Posts { get; set; }
}

public class Post
{
    public int Id { get; set; }
    public string Title { get; set; }
    public Blog Blog { get; set; }
}
```

</details>

### PR DESCRIPTION
**Pull request by maumar**
When processing group by we use all the grouping properties as identifiers (so that we know how to bucket the results). For final groupby we convert the group by into orderby. In both of those cases, if grouping key is entity type, we don't need to use all the properties. Only using PK properties is sufficient to uniquely identify/order the results.

Fixes #34895

### CODE REVIEW

File: test/EFCore.SqlServer.FunctionalTests/Query/NorthwindGroupByQuerySqlServerTest.cs [hunk 1]
```diff
@@ -3614,12 +3614,12 @@ public override async Task Final_GroupBy_entity(bool async)
         await base.Final_GroupBy_entity(async);
 
         AssertSql(
-            """
+"""
```
Comment (roji) [inline test/EFCore.SqlServer.FunctionalTests/Query/NorthwindGroupByQuerySqlServerTest.cs:3617]: nit: revert

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 2]
```diff
@@ -527,9 +527,24 @@ public Expression ApplyProjection(
             }
 
             // Convert GroupBy to OrderBy
-            foreach (var groupingTerm in _groupBy)
```
Comment (roji) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:530]: @maumar I'll take a deeper look tomorrow (it's late), but I'm wondering: how come `_groupBy` contains all the properties in the first place? In other words, shouldn't we be fixing this further up the stream, at the point where we're populating `_groupBy` itself, rather than introducing a special case and bypassing it here? Fixing it "at the source" may improve other cases too.

Then again, I may be misunderstanding things completely, since I'm not yet very familiar with the GroupBy implementation... I'm hoping/planning to put some time into working on GroupBy for 10, so that can also be part of that (but of course am more than happy if you want to pursue here)

Comment (maumar) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:530]: _groupBy is populated in the ApplyGrouping -> PopulateGroupByTerms. That logic is shared for all group bys, before we know if its final group by or not. I think we need all the properties here, because we don't know what's composed on top. E.g. I can group by Customer, and later on just project only FirstName. If we only hold customer Id, we can't do that anymore. Will add a test like that, to make sure my understanding is sound.

Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:530]: >  That logic is shared for all group bys, before we know if its final group by or not. 

Of course - but it seems wrong for `_groupBy` to contain all the properties regardless of whether the GroupBy is final or not. I did a quick test, and the problem indeed isn't constrained to final GroupBy; consider the following:

```c#
_ = await context.Posts
    .GroupBy(p => p.Blog)
    .Select(g => new { g.Key.Id, Count = g.Count() })
    .ToListAsync();
```

This translates to the following:

```sql
SELECT [b].[Id], COUNT(*) AS [Count]
FROM [Posts] AS [p]
INNER JOIN [Blogs] AS [b] ON [p].[BlogId] = [b].[Id]
GROUP BY [b].[Id], [b].[Name]
```

Note that the LINQ query doesn't project the Blog's Name out, and there's no reason for us to group by Name (grouping by the primary key should be enough - just like in the final GroupBy case). But our SQL tells the database to do the (considerable!) extra work of grouping by Name.

In other words, this seems like a larger issue that needs to be fixed at the root, rather than having a final-GroupBy-specific thing, no?

Comment (maumar) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:530]: added a test that would have failed if _groupBy only contained entity PKs

Comment (maumar) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:530]: ahh I see, problem is we only know later if the non-PK will be part of a projection, i.e. this case:

```cs
_ = await context.Posts
    .GroupBy(p => p.Blog)
    .Select(g => new { g.Key.Name, Count = g.Count() })
    .ToListAsync();
```

We would need to do some sort of pruning once we know the projection, but that's tricky

Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:530]: My initial unfounded expectation here would be to only add something to the grouping terms once they get projected out, rather than adding everything initially, and then have to worry about pruning them later or similar... Our intermediate state here (before the select is applied) simply seems wrong, no?

But of course anything here would require more thinking/investigation...

Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:530]: A couple more thoughts...

First, trying to solve this via pruning is indeed tricky, as you say; it means that for a grouping key, we need to be able to figure out if a subset of the columns constitute a unique set (e.g. a primary key), and at that point we can prune any other column - assuming it's not referenced (e.g. projected). This isn't easy, `x.Id` is unique, `x.Id + 1` is also unique, but `x.Id % 2` is not, etc. I'm not sure we should go in this direction...

More philosophically, this problem generally stems from our design, where we want to be able to just bind a property and generate a ColumnExpression, without that affecting the SelectExpression; or more generally, it's our (flawed) design where RelationalSqlTranslatingEV is expected to be able to accept any lambda and return its translation, without any state or modification to the current SelectExpression. In other words, I think the right design here is for our translation of GroupBy() to only popualate the actual columns specified in the operator (e.g. if it's an entity type, just put the primary key columns). Then, if a non-PK column needs to be projected (or otherwise accessed), we should update SelectExpression.GroupBy at that point, as part of binding the property (again, this requires modifying the SelectExpression from within RelationalSqlTranslatingEV/ProjectionBindingEV, which is the thing that's against our current design).

This is a bit similar to our approach with nav expansion. Instead of adding JOINs at the point where a property is bound via the navigation, we preprocess the tree and transform it, so that when we reach translation, the tree has been remade in a way that JOINs have already been added and properties can just be bound without adding JOINs. Removing the nav expansion preprocessing step would mean adding the JOIN during translation, at the moment where a property is bound - similar to how we'd add the GROUP BY column when the corresponding property is bound.
